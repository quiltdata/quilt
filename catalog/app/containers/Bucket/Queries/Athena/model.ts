export const Loading = Symbol('loading')

export type Maybe<T> = T | null

// `T` is the value
// `undefined` is no value. It is not loaded or doesn't exist in loaded data
// `Loading` is loading, ok?
// `Error` is error, ok?
export type Data<T> = T | undefined | typeof Loading | Error

export interface DataController<T> {
  data: Data<T>
  loadMore: () => void
}

// `T` is the value
// `null` is no value, explicitly set by user
// `undefined` is no value. It is not loaded or doesn't exist in loaded data
// `Loading` is loading, ok?
// `Error` is error, ok?
export type Value<T> = Maybe<Data<T>>

export interface ValueController<T> {
  value: Value<T>
  setValue: (v: T | null) => void
}

// Bad

export function isFulfilled<T>(value: Value<T>): value is T | null {
  if (value === undefined || value === Loading || value instanceof Error) {
    return false
  }
  return true
}

// Data is loaded
// or value is set to actual value
export function hasData<T>(value: Value<T>): value is T {
  if (
    value === undefined ||
    value === Loading ||
    value instanceof Error ||
    value === null
  ) {
    return false
  }
  return true
}

// Value is selected with some or no value, or resolved with error
// Or, data is loaded
export function isObtained<T>(value: Value<T>): value is T | null | Error {
  if (value === undefined || value === Loading) {
    return false
  }
  return true
}

// Value is selected with some or no value
// Or, data is loaded successfully
export function hasValue<T>(value: Value<T>): value is T | null {
  if (value === undefined || value === Loading || value instanceof Error) {
    return false
  }
  return true
}

// Good

export function isSelected<T>(value: Value<T>): value is T {
  if (
    value === undefined ||
    value === Loading ||
    value instanceof Error ||
    value === null
  ) {
    return false
  }
  return true
}

export function isError<T>(value: Value<T>): value is Error {
  return value instanceof Error
}

// Data is loading, or value is waiting for data
export function isLoading<T>(value: Value<T>): value is typeof Loading {
  return value === Loading
}

// No value yet, value or data was just initialized
export function isNone<T>(value: Value<T>): value is undefined {
  return value === undefined
}

// User explicitly set no value
export function isNoneSelected<T>(value: Value<T>): value is null {
  return value === null
}
