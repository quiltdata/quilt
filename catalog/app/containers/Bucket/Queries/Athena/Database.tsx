import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'
import * as Lab from '@material-ui/lab'

import Skeleton from 'components/Skeleton'
import * as Dialogs from 'utils/GlobalDialogs'

import * as Model from './model'
import * as storage from './model/storage'

const useSelectErrorStyles = M.makeStyles((t) => ({
  button: {
    whiteSpace: 'nowrap',
  },
  dialog: {
    padding: t.spacing(2),
  },
}))

interface SelectErrorProps {
  className?: string
  error: Error
}

function SelectError({ className, error }: SelectErrorProps) {
  const classes = useSelectErrorStyles()
  const openDialog = Dialogs.use()
  const handleClick = React.useCallback(() => {
    openDialog(() => (
      <div className={classes.dialog}>
        <M.Typography>{error.message}</M.Typography>
      </div>
    ))
  }, [classes.dialog, error.message, openDialog])
  return (
    <Lab.Alert
      action={
        <M.Button
          className={classes.button}
          color="inherit"
          onClick={handleClick}
          size="small"
        >
          Show more
        </M.Button>
      }
      className={className}
      severity="error"
    >
      {error.name}
    </Lab.Alert>
  )
}

const LOAD_MORE = '__load-more__'

const EMPTY = '__empty__'

interface Response {
  list: string[]
  next?: string
}

const useSelectStyles = M.makeStyles({
  root: {
    width: '100%',
  },
})

interface SelectProps {
  className?: string
  data: Response
  label: string
  onChange: (value: string) => void
  onLoadMore: (prev: Response) => void
  value: string | null
  disabled?: boolean
}

function Select({
  className,
  data,
  disabled,
  label,
  onChange,
  onLoadMore,
  value,
}: SelectProps) {
  const classes = useSelectStyles()
  const handleChange = React.useCallback(
    (event) => {
      if (event.target.value === LOAD_MORE) {
        onLoadMore(data)
      } else {
        onChange(event.target.value)
      }
    },
    [data, onLoadMore, onChange],
  )

  return (
    <M.FormControl className={cx(classes.root, className)} disabled={disabled}>
      <M.InputLabel>{label}</M.InputLabel>
      <M.Select
        onChange={handleChange}
        value={data.list.length ? value?.toLowerCase() || '' : EMPTY}
        disabled={disabled || !data.list.length}
      >
        {data.list.map((item) => (
          <M.MenuItem key={item} value={item.toLowerCase()}>
            {item}
          </M.MenuItem>
        ))}
        {data.next && <M.MenuItem value={LOAD_MORE}>Load more</M.MenuItem>}
        {!data.list.length && (
          <M.MenuItem value={value?.toLowerCase() || EMPTY}>
            {value || 'Empty list'}
          </M.MenuItem>
        )}
      </M.Select>
    </M.FormControl>
  )
}

interface SelectCatalogNameProps {
  className?: string
}

function SelectCatalogName({ className }: SelectCatalogNameProps) {
  const { catalogName, catalogNames, queryRun } = Model.use()

  const handleChange = React.useCallback(
    (value) => {
      storage.setCatalog(value)
      storage.clearDatabase()
      catalogName.setValue(value)
    },
    [catalogName],
  )

  if (Model.isError(catalogNames.data)) {
    return <SelectError className={className} error={catalogNames.data} />
  }
  if (Model.isError(catalogName.value)) {
    return <SelectError className={className} error={catalogName.value} />
  }
  if (!Model.hasValue(catalogName.value) || !Model.hasData(catalogNames.data)) {
    return <Skeleton className={className} height={32} animate mt={2} />
  }

  return (
    <Select
      className={className}
      data={catalogNames.data}
      disabled={Model.isLoading(queryRun)}
      label="Data catalog"
      onChange={handleChange}
      onLoadMore={catalogNames.loadMore}
      value={catalogName.value}
    />
  )
}

interface SelectDatabaseProps {
  className: string
}

function SelectDatabase({ className }: SelectDatabaseProps) {
  const { catalogName, database, databases, queryRun } = Model.use()

  const handleChange = React.useCallback(
    (value) => {
      storage.setDatabase(value)
      database.setValue(value)
    },
    [database],
  )

  if (Model.isError(databases.data)) {
    return <SelectError className={className} error={databases.data} />
  }
  if (Model.isError(database.value)) {
    return <SelectError className={className} error={database.value} />
  }
  if (!Model.hasValue(database.value) || !Model.hasData(databases.data)) {
    return <Skeleton className={className} height={32} animate mt={2} />
  }

  return (
    <Select
      data={databases.data}
      disabled={!Model.hasValue(catalogName) || Model.isLoading(queryRun)}
      label="Database"
      onChange={handleChange}
      onLoadMore={databases.loadMore}
      value={database.value}
    />
  )
}

const useStyles = M.makeStyles((t) => ({
  root: {
    alignItems: 'center',
    display: 'flex',
  },
  field: {
    flexBasis: '50%',
    marginRight: t.spacing(2),
  },
  button: {
    marginLeft: t.spacing(1),
  },
}))

interface DatabaseProps {
  className?: string
}

export default function Database({ className }: DatabaseProps) {
  const classes = useStyles()
  return (
    <div className={cx(classes.root, className)}>
      <SelectCatalogName className={classes.field} />
      <SelectDatabase className={classes.field} />
    </div>
  )
}
