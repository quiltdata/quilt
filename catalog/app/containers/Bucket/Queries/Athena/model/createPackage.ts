import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

import Log from 'utils/Logging'
import type * as requests from './requests'

export function doQueryResultsContainManifestEntries(
  queryResults: requests.QueryResults,
): queryResults is requests.QueryManifests {
  if (!queryResults.rows.length) return false
  const columnNames = queryResults.columns.map(({ name }) => name)
  return (
    columnNames.includes('size') &&
    (columnNames.includes('physical_keys') || columnNames.includes('physical_key')) &&
    columnNames.includes('logical_key')
  )
}

type Row = requests.QueryManifests['rows'][0]
function parseRow(
  row: Row,
  columns: requests.QueryResultsColumns,
): { fail?: undefined; ok: [string, Model.S3File] } | { fail: Row; ok?: undefined } {
  try {
    const entry = row.reduce(
      (acc, value, index) => {
        if (!columns[index]?.name) return acc
        return {
          ...acc,
          [columns[index].name]: value,
        }
      },
      {} as Record<requests.ManifestKey, string>,
    )
    if (!entry.logical_key) return { fail: row }
    if (!entry.physical_key && !entry.physical_keys) return { fail: row }
    const handle = entry.physical_key
      ? s3paths.parseS3Url(entry.physical_key)
      : s3paths.parseS3Url(entry.physical_keys.replace(/^\[/, '').replace(/\]$/, ''))
    const sizeParsed = Number(entry.size)
    const size = Number.isNaN(sizeParsed) ? 0 : sizeParsed
    return {
      ok: [
        entry.logical_key,
        {
          ...handle,
          size,
        },
      ],
    }
  } catch (e) {
    Log.error(e)
    return { fail: row }
  }
}

export interface ParsedRows {
  valid: Record<string, Model.S3File>
  invalid: requests.QueryResultsRows
}

export function parseQueryResults(queryResults: requests.QueryManifests): ParsedRows {
  return queryResults.rows
    .map((row) => parseRow(row, queryResults.columns))
    .reduce(
      (memo, { ok, fail }) =>
        ok
          ? {
              valid: {
                ...memo.valid,
                [ok[0]]: ok[1],
              },
              invalid: memo.invalid,
            }
          : {
              valid: memo.valid,
              invalid: [...memo.invalid, fail],
            },
      { valid: {}, invalid: [] } as ParsedRows,
    )
}
