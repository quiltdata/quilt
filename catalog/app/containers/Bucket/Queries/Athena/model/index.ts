export type * from './requests'
export { NO_DATABASE } from './requests'
export * from './state'
export * from './utils'
