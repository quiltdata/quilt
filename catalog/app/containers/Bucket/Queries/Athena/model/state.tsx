import invariant from 'invariant'
import * as React from 'react'
import * as RRDom from 'react-router-dom'

import type * as BucketPreferences from 'utils/BucketPreferences'
import * as NamedRoutes from 'utils/NamedRoutes'

import * as requests from './requests'
import * as Model from './utils'

export interface State {
  bucket: string
  queryExecutionId?: string

  /**
   * Query execution loaded by id on the corresponding page.
   * On the index page (where there is no queryExecutionId) its value is null.
   */
  execution: Model.Value<requests.QueryExecution>

  /** List of workgroups from Athena */
  workgroups: Model.DataController<Model.List<requests.Workgroup>>
  /**
   * Workgroup selected by user explicitly or from page URL, and validated that it does exist
   * If workgroup doesn't exist, then its value is Error
   * It can't be null
   */
  workgroup: Model.DataController<requests.Workgroup>
  /** List of named queries, including query body for each query */
  queries: Model.DataController<Model.List<requests.Query>>
  /** Selected named query */
  query: Model.ValueController<requests.Query>
  /** Query body, typed by user or set from selected named query or query execution */
  queryBody: Model.ValueController<string>
  /** List of catalog names from Athena */
  catalogNames: Model.DataController<Model.List<requests.CatalogName>>
  /** Catalog name selected by user, or set initially */
  catalogName: Model.ValueController<requests.CatalogName>
  /** List of databases from Athena */
  databases: Model.DataController<Model.List<requests.Database>>
  /** Database selected by user, or set initially */
  database: Model.ValueController<requests.Database>
  /** List of query executions, in other words, history of executions */
  executions: Model.DataController<Model.List<requests.QueryExecutionsItem>>
  /** Rows and columns of query results */
  results: Model.DataController<requests.QueryResults>

  /**
   * Submit query to Athena with values memoized here in state
   * If catalog name or database is not selected, then it will return specific output
   * Which is handled and then user can re-submit with `forceDefaultExecutionContext: true`
   */
  submit: (
    forceDefaultExecutionContext: boolean,
  ) => Promise<Model.Value<requests.QueryRun>>
  /**
   * Query run is `undefined` when there is not enough data to run the query
   * It is `null` when it is ready to run
   * Error when submit failed or when validation failed (e.g. no database selected)
   */
  queryRun: Model.Value<requests.QueryRun>
}

export const Ctx = React.createContext<State | null>(null)

interface ProviderProps {
  preferences?: BucketPreferences.AthenaPreferences
  children: React.ReactNode
}

export function Provider({ preferences, children }: ProviderProps) {
  const { urls } = NamedRoutes.use()

  const {
    bucket,
    queryExecutionId,
    workgroup: workgroupId,
  } = RRDom.useParams<{
    bucket: string
    queryExecutionId?: string
    workgroup?: requests.Workgroup
  }>()
  invariant(!!bucket, '`bucket` must be defined')

  const execution = requests.useWaitForQueryExecution(queryExecutionId)

  const workgroups = requests.useWorkgroups()
  const workgroup = requests.useWorkgroup(workgroups, workgroupId, preferences)
  const queries = requests.useQueries(workgroup.data)
  const query = requests.useQuery(queries.data, execution)
  const queryBody = requests.useQueryBody(query.value, query.setValue, execution)
  const catalogNames = requests.useCatalogNames(workgroup.data)
  const catalogName = requests.useCatalogName(catalogNames.data, execution)
  const databases = requests.useDatabases(catalogName.value)
  const database = requests.useDatabase(databases.data, execution)
  const executions = requests.useExecutions(workgroup.data, queryExecutionId)
  const results = requests.useResults(execution)

  const [queryRun, submit] = requests.useQueryRun({
    workgroup: workgroup.data,
    catalogName: catalogName.value,
    database: database.value,
    queryBody: queryBody.value,
  })

  const value: State = {
    bucket,
    queryExecutionId,
    workgroup,

    catalogName,
    catalogNames,
    database,
    databases,
    execution,
    executions,
    queries,
    query,
    queryBody,
    results,
    workgroups,

    submit,
    queryRun,
  }

  if (Model.hasData(queryRun) && queryExecutionId !== queryRun.id) {
    return (
      <RRDom.Redirect
        to={urls.bucketAthenaExecution(bucket, workgroup.data, queryRun.id)}
      />
    )
  }

  if (Model.hasData(workgroup.data) && !workgroupId) {
    return <RRDom.Redirect to={urls.bucketAthenaWorkgroup(bucket, workgroup.data)} />
  }

  return <Ctx.Provider value={value}>{children}</Ctx.Provider>
}

/** state object is not memoized, use destructuring down to memoized properties */
export function useState() {
  const model = React.useContext(Ctx)
  invariant(model, 'Athena state accessed outside of provider')
  return model
}

export const use = useState
