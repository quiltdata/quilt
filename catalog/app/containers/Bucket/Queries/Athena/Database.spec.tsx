import * as React from 'react'
import renderer from 'react-test-renderer'

import WithGlobalDialogs from 'utils/GlobalDialogs'

import Database from './Database'

import * as Model from './model'

jest.mock(
  'constants/config',
  jest.fn(() => ({})),
)

const noop = () => {}

const emptyState: Model.State = {
  bucket: 'any',

  catalogName: { value: undefined, setValue: noop },
  catalogNames: { data: undefined, loadMore: noop },
  database: { value: undefined, setValue: noop },
  databases: { data: undefined, loadMore: noop },
  execution: undefined,
  executions: { data: undefined, loadMore: noop },
  queries: { data: undefined, loadMore: noop },
  query: { value: undefined, setValue: noop },
  queryBody: { value: undefined, setValue: noop },
  results: { data: undefined, loadMore: noop },
  workgroups: { data: undefined, loadMore: noop },
  workgroup: { data: undefined, loadMore: noop },

  submit: () => Promise.resolve({ id: 'bar' }),
  queryRun: undefined,
}

interface ProviderProps {
  children: React.ReactNode
  value: Model.State
}

function Provider({ children, value }: ProviderProps) {
  return <Model.Ctx.Provider value={value}>{children}</Model.Ctx.Provider>
}

describe('containers/Bucket/Queries/Athena/Database', () => {
  beforeAll(() => {})

  afterAll(() => {})

  it('should render skeletons', () => {
    const tree = renderer.create(
      <Provider value={emptyState}>
        <Database />
      </Provider>,
    )
    expect(tree).toMatchSnapshot()
  })

  it('should render selected values', () => {
    const tree = renderer.create(
      <Provider
        value={{
          ...emptyState,
          catalogName: Model.wrapValue('foo', noop),
          catalogNames: Model.wrapData({ list: ['foo'] }, noop),
          databases: Model.wrapData({ list: ['bar'] }, noop),
          database: Model.wrapValue('bar', noop),
        }}
      >
        <Database />
      </Provider>,
    )
    expect(tree).toMatchSnapshot()
  })

  it('should show no value (zero-width space) if selected no value', () => {
    const tree = renderer.create(
      <Provider
        value={{
          ...emptyState,
          catalogName: { value: null, setValue: noop },
          catalogNames: Model.wrapData({ list: ['any'] }, noop),
          databases: Model.wrapData({ list: ['any'] }, noop),
          database: { value: null, setValue: noop },
        }}
      >
        <Database />
      </Provider>,
    )
    expect(tree).toMatchSnapshot()
  })

  it('should disable selection if no spare values', () => {
    const tree = renderer.create(
      <Provider
        value={{
          ...emptyState,
          catalogName: { value: null, setValue: noop },
          catalogNames: Model.wrapData({ list: [] }, noop),
          databases: Model.wrapData({ list: [] }, noop),
          database: { value: null, setValue: noop },
        }}
      >
        <Database />
      </Provider>,
    )
    expect(tree).toMatchSnapshot()
  })

  it('should show error when values failed', () => {
    const tree = renderer.create(
      <WithGlobalDialogs>
        <Provider
          value={{
            ...emptyState,
            catalogName: { value: new Error('Value fail'), setValue: noop },
            database: { value: new Error('Value fail'), setValue: noop },
          }}
        >
          <Database />
        </Provider>
      </WithGlobalDialogs>,
    )
    expect(tree).toMatchSnapshot()
  })

  it('should show error when data failed', () => {
    const tree = renderer.create(
      <WithGlobalDialogs>
        <Provider
          value={{
            ...emptyState,
            catalogNames: { data: new Error('Data fail'), loadMore: noop },
            databases: { data: new Error('Data fail'), loadMore: noop },
          }}
        >
          <Database />
        </Provider>
      </WithGlobalDialogs>,
    )
    expect(tree).toMatchSnapshot()
  })
})
