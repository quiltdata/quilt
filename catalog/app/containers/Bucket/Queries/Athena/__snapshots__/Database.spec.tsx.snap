// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Bucket/Queries/Athena/Database should disable selection if no spare values 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiFormControl-root makeStyles-root-9 makeStyles-field-2"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink={true}
    >
      Data catalog
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-disabled="true"
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled Mui-disabled"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={null}
        role="button"
        tabIndex={null}
      >
        Empty list
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value="__empty__"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon Mui-disabled"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiFormControl-root makeStyles-root-9"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink={true}
    >
      Database
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-disabled="true"
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled Mui-disabled"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={null}
        role="button"
        tabIndex={null}
      >
        Empty list
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value="__empty__"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon Mui-disabled"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`containers/Bucket/Queries/Athena/Database should render selected values 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiFormControl-root makeStyles-root-9 makeStyles-field-2"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink={true}
    >
      Data catalog
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        foo
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value="foo"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiFormControl-root makeStyles-root-9"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink={true}
    >
      Database
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        bar
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value="bar"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`containers/Bucket/Queries/Athena/Database should render skeletons 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiBox-root MuiBox-root-7 makeStyles-field-2 makeStyles-root-4 makeStyles-animate-6"
  />
  <div
    className="MuiBox-root MuiBox-root-8 makeStyles-field-2 makeStyles-root-4 makeStyles-animate-6"
  />
</div>
`;

exports[`containers/Bucket/Queries/Athena/Database should show error when data failed 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiPaper-root MuiAlert-root MuiAlert-standardError makeStyles-field-2 MuiPaper-elevation0"
    role="alert"
  >
    <div
      className="MuiAlert-icon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      className="MuiAlert-message"
    >
      Error
    </div>
    <div
      className="MuiAlert-action"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-10 MuiButton-colorInherit MuiButton-textSizeSmall MuiButton-sizeSmall"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Show more
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    className="MuiPaper-root MuiAlert-root MuiAlert-standardError makeStyles-field-2 MuiPaper-elevation0"
    role="alert"
  >
    <div
      className="MuiAlert-icon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      className="MuiAlert-message"
    >
      Error
    </div>
    <div
      className="MuiAlert-action"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-10 MuiButton-colorInherit MuiButton-textSizeSmall MuiButton-sizeSmall"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Show more
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`containers/Bucket/Queries/Athena/Database should show error when values failed 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiPaper-root MuiAlert-root MuiAlert-standardError makeStyles-field-2 MuiPaper-elevation0"
    role="alert"
  >
    <div
      className="MuiAlert-icon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      className="MuiAlert-message"
    >
      Error
    </div>
    <div
      className="MuiAlert-action"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-10 MuiButton-colorInherit MuiButton-textSizeSmall MuiButton-sizeSmall"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Show more
        </span>
      </button>
    </div>
  </div>
  <div
    className="MuiPaper-root MuiAlert-root MuiAlert-standardError makeStyles-field-2 MuiPaper-elevation0"
    role="alert"
  >
    <div
      className="MuiAlert-icon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      className="MuiAlert-message"
    >
      Error
    </div>
    <div
      className="MuiAlert-action"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-10 MuiButton-colorInherit MuiButton-textSizeSmall MuiButton-sizeSmall"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Show more
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`containers/Bucket/Queries/Athena/Database should show no value (zero-width space) if selected no value 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiFormControl-root makeStyles-root-9 makeStyles-field-2"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
    >
      Data catalog
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            {
              "__html": "&#8203;",
            }
          }
        />
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value=""
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiFormControl-root makeStyles-root-9"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
    >
      Database
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            {
              "__html": "&#8203;",
            }
          }
        />
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value=""
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
